mautic/mautic
# [UX] Filter choices for field order when creating Custom Fields

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 
| New feature/enhancement? (use the a.x branch)      | 🟢 - enhancement
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🔴 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #14180 & #14181 



## Description
This PR solves issues #14180 & #14181 with some additional related changes
- Updated order field query_builder so that choices are filtered by selected object and group
- Added Ajax so that the choice list is updated when object or group changes, which involved refactoring how the field is added
- Disabled Field order of Fixed field and set tooltip accordingly
- Order the fields in quick add form so that Fixed fields come first

Note the reorderFieldsByEntity method might be buggy, and sometimes the fields are placed below instead of above selected field. Not sure what is causing it and might need some guidance if it needs to be fixed in this PR.






---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Go to custom fields and add a couple of custom fields
3. See that the options in field order changes based on the object and group selected
4. Once you've added  custom fields, check the form by going to contacts/companies and clicking on new
5. Edit a newly created field and see that you can change the group and the order
6. Go to existing fields that cannot be reordered and see that the field order is disabled and the tooltip states that default field cannot be reordered
7. Create a new field where "Visible on quick add" is checked and see that the field is placed below the fixed fields such as first name and last name
